chainer/gradient_check.py

```diff
@@ -555,18 +555,45 @@ def _sample_directions(self):
             None if shape is None
             else xp.random.normal(size=shape) for shape in direction_shapes]
 
+        if all([d is None or d.size == 0 for d in directions]):
+            return directions
+
+        # For simplicity omit None from the vector. They're recovered later.
+        none_indices = [i for i, d in enumerate(directions) if d is None]
+        directions = [d for d in directions if d is not None]
+
         # The direction vector is normalized in order to keep the scale of
         # differentiation error invariant with respect to the number of input
         # dimensions. Ideally, the scale of the curvature with respect to each
         # input dimension should be taken into account, but we ignore the
         # differences and assume that the curvature is uniform with respect to
         # all the input dimensions.
-        norm = math.sqrt(
-            sum([0 if d is None else xp.square(d).sum() for d in directions]))
-        if norm != 0:
-            # norm could be zero if input arrays are 0-sized.
-            scale = 1. / norm
-            directions = [None if d is None else d * scale for d in directions]
+        norm = math.sqrt(sum([xp.square(d).sum() for d in directions]))
+        scale = 1. / norm
+        directions = [xp.asarray(d * scale) for d in directions]
+
+        # Small elements in the direction vector leads to instability on
+        # gradients comparison.
+        # In order to avoid that, absolute values are capped at 0.1 / sqrt(N)
+        # at minimum, where N is the number of elements. Other elements are
+        # scaled uniformly so that the total L2 norm will remain 1.
+        min_d = 0.1 / math.sqrt(sum([d.size for d in directions]))
+        is_small = [min_d > xp.abs(d) for d in directions]
+        is_large = [xp.logical_not(iss) for iss in is_small]
+        n_small = sum([iss.sum() for iss in is_small])
+        sq_large = sum(
+            [xp.square(d[isl]).sum() for d, isl in zip(directions, is_large)])
+        scale = xp.sqrt((1 - n_small * min_d ** 2) / sq_large)
+
+        for d, iss, isl in zip(directions, is_small, is_large):
+            # Cap small elements.
+            d[iss] = xp.sign(d[iss]) * min_d
+            # Scale large elements.
+            d[isl] *= scale
```

Comment:
- Scaled values are not always capped at 0.1 / sqrt(N).
- Using the same value at many indices might cause issue.

To address these, I suggest the following algorithm
```
>>> d = np.random.normal(size=(20, 30))
>>> d_sq = np.square(d)
>>> d_new = np.copysign(np.sqrt((0.99 / d_sq.sum()) * d_sq + 0.01 / d.size), d)
>>> np.square(d_new).sum()
1.0
>>> np.sort(abs(d_new.ravel()))[:10] - 0.1 / np.sqrt(20 * 30)
array([2.8439929906868416e-06, 5.6006727285236313e-06,
       1.4866622380856585e-05, 2.1999054026286893e-05,
       4.8911109011619616e-05, 5.7554843792876308e-05,
       6.0676131907084173e-05, 8.1905077464757003e-05,
       1.1672753697054838e-04, 1.5592912466356589e-04])
>>> np.sort(abs(d_new.ravel()))[-10:]
array([0.09640808766675514, 0.097591641176575  , 0.10326711220713523,
       0.10351729954309667, 0.10741205037687222, 0.10755555726558051,
       0.10905204794060573, 0.11080711821818465, 0.1121366033473858 ,
       0.12318624894785492])
```

Reply:
What's the idea behind this formula? I can't immediately undestand it.

Reply:
- Its abs is a weighted root mean square of abs of `d / norm(d)` and `full(d.shape, 1.0 / sqrt(d.size))`.
- Its sign is the sign of `d / norm(d)`.